pkg/merkletree/tree_test.go

```diff
@@ -0,0 +1,357 @@
+
+	t.Run("no leaves", func(t *testing.T) {
+		t.Parallel()
+		var err error
+
+		_, err = NewTree([][]byte{}, nil)
+		require.Error(t, err)
+
+		_, err = NewTree(nil, nil)
+		require.Error(t, err)
+	})
+}
+
+func TestGetProof(t *testing.T) {
+	t.Parallel()
+	t.Run("five nodes", func(t *testing.T) {
+		t.Parallel()
+		leaves := [][]byte{
+			[]byte("001"),
+			[]byte("002"),
+			[]byte("003"),
+			[]byte("004"),
+			[]byte("005"),
+		}
+
+		tree, err := NewTree(leaves, mockHashFunc)
+		require.NoError(t, err)
+		require.Equal(t, [][]byte{[]byte("005"), []byte("010"), []byte("010")}, tree.GetProof([]byte("005")))
+	})
+
+	t.Run("eight nodes", func(t *testing.T) {
+		t.Parallel()
+		leaves := [][]byte{
+			[]byte("001"),
+			[]byte("002"),
+			[]byte("003"),
+			[]byte("004"),
+			[]byte("005"),
+			[]byte("006"),
+			[]byte("007"),
+			[]byte("008"),
+		}
+
+		tree, err := NewTree(leaves, mockHashFunc)
+		require.NoError(t, err)
+		require.Equal(t, [][]byte{[]byte("002"), []byte("007"), []byte("026")}, tree.GetProof([]byte("001")))
+		require.Equal(t, [][]byte{[]byte("001"), []byte("007"), []byte("026")}, tree.GetProof([]byte("002")))
+		require.Equal(t, [][]byte{[]byte("004"), []byte("003"), []byte("026")}, tree.GetProof([]byte("003")))
+		require.Equal(t, [][]byte{[]byte("003"), []byte("003"), []byte("026")}, tree.GetProof([]byte("004")))
+
+		require.Equal(t, [][]byte{[]byte("006"), []byte("015"), []byte("010")}, tree.GetProof([]byte("005")))
+		require.Equal(t, [][]byte{[]byte("005"), []byte("015"), []byte("010")}, tree.GetProof([]byte("006")))
+		require.Equal(t, [][]byte{[]byte("008"), []byte("011"), []byte("010")}, tree.GetProof([]byte("007")))
+		require.Equal(t, [][]byte{[]byte("007"), []byte("011"), []byte("010")}, tree.GetProof([]byte("008")))
+	})
+
+	t.Run("not found", func(t *testing.T) {
+		t.Parallel()
+		leaves := [][]byte{
+			[]byte("001"),
+			[]byte("002"),
+			[]byte("003"),
+			[]byte("004"),
+			[]byte("005"),
+		}
+
+		tree, err := NewTree(leaves, mockHashFunc)
+		require.NoError(t, err)
+
+		require.Nil(t, tree.GetProof([]byte("006")))
+		require.Len(t, tree.GetProof([]byte("006")), 0)
+	})
+}
+
+func TestVerifyProof(t *testing.T) {
+	t.Parallel()
+	t.Run("correct proof", func(t *testing.T) {
+		t.Parallel()
+		root := []byte("036")
+		proof := [][]byte{[]byte("002"), []byte("007"), []byte("026")}
+		leaf := []byte("001")
+		require.True(t, VerifyProof(root, proof, leaf, mockHashFunc))
+	})
+
+	t.Run("wrong root", func(t *testing.T) {
+		t.Parallel()
+		root := []byte("035")
+		proof := [][]byte{[]byte("002"), []byte("007"), []byte("026")}
+		leaf := []byte("001")
+		require.False(t, VerifyProof(root, proof, leaf, mockHashFunc))
+	})
+
+	t.Run("wrong proof", func(t *testing.T) {
+		t.Parallel()
+		root := []byte("036")
+		proof := [][]byte{[]byte("001"), []byte("007"), []byte("026")}
+		leaf := []byte("001")
+		require.False(t, VerifyProof(root, proof, leaf, mockHashFunc))
+	})
+}
+
+func TestProperties(t *testing.T) {
```

Comment:
Experimenting with the `testing/quick` library to test properties of the tree and catch problems. It was quite good actually, found 2 bugs in my first implementation.